cli/src/pcluster/models/imagebuilder.py

```diff
@@ -146,6 +167,89 @@ def config(self):
             self.__config = ImageBuilderSchema().load(self.source_config)
         return self.__config
 
+    @property
+    def bucket(self):
+        """Return a bucket configuration."""
+        if self.__bucket:
+            return self.__bucket
+
+        if self.__source_config:
+            custom_bucket_name = self.config.custom_s3_bucket
+        else:
+            custom_bucket_name = self._get_custom_bucket()
+
+        self.__bucket = S3BucketFactory.init_s3_bucket(
+            service_name=self.image_name,
+            stack_name=self.image_name,
+            custom_s3_bucket=custom_bucket_name,
+            artifact_directory=self.s3_artifacts_dir,
+        )
+        return self.__bucket
+
+    def _get_custom_bucket(self):
+        """Try to get custom bucket name from image tag or stack tag."""
+        try:
+            image_info = AWSApi.instance().ec2.describe_image_by_name_tag(self.image_name)
+            custom_bucket_name = search_tag(image_info, PCLUSTER_S3_BUCKET_TAG)
+            if custom_bucket_name == S3Bucket.get_bucket_name(AWSApi.instance().sts.get_account_id(), get_region()):
+                custom_bucket_name = None
+        except AWSClientError as e:
+            if not isinstance(e, ImageNotFoundError):
+                raise ImageBuilderActionError(f"Unable to get S3 bucket name from image {self.image_name} tag. {e}")
+            custom_bucket_name = None
+
+        if custom_bucket_name is None:
+            try:
+                custom_bucket_name = self.stack.s3_bucket_name
+                if custom_bucket_name == S3Bucket.get_bucket_name(AWSApi.instance().sts.get_account_id(), get_region()):
+                    custom_bucket_name = None
+            except AWSClientError as e:
+                raise ImageBuilderActionError(f"Unable to get S3 bucket name from stack {self.stack.name} tag. {e}")
+
+        return custom_bucket_name
+
+    def _get_artifact_dir(self):
+        """Get artifact directory from image tag or stack tag."""
+        try:
+            image_info = AWSApi.instance().ec2.describe_image_by_name_tag(self.image_name)
```

Comment:
we can optimize this in a separate PR but describe_image_by_name_tag is called multiple times while it could be reused (e.g. _get_custom_bucket is calling it)

Reply:
Done

Reply:
You can even add Cached annotation to describe_image_by_name_tag call in AWS Api layer.

Reply:
I prefer using an ImageInfo class to store the describe image result like StackInfo, and add a property in ImageBuilder, it has been implemented in pcluster delete PR. https://github.com/aws/aws-parallelcluster/pull/2507